### DISCUSSION TITLE
Why or when is Equatable of Action needed?

### DISCUSSION THREAD

**Discussion by hmhv** [pointfreeco/swift-composable-architecture#1740] [Q&A] [answered]
Hi~. I am new to TCA.
I have read the ReadME and have a question.
Why or when is Equatable of Action needed?
I found that Equatable of State is required when using State within ViewState.

```swift
struct Feature: ReducerProtocol {
  struct State: Equatable { … }
  enum Action: Equatable { … }
...
}
```

**Comment by tgrapperon** [comment, marked as answer]
Hey @hmhv! In both cases, `Equatable` is used by `TestStore`, as it needs to be able to compare actions produced in `Effect`s, as well a states before and after reductions.

Please note that an `Equatable` `Reducer.State` is not a requirement for `ViewState`. It's even sometimes an undesirable side effect of testability, as its immediate availability as `Equatable` encourages you to observe it directly in a `ViewStore` when a dedicated (and `Equatable`) `ViewState` may be preferable.

**Reply by hmhv** [reply]
@tgrapperon 
Thank you!

---
I found the lines from `TestStore.swift` and understood why `send` and `receive` methods need `Equatable`.

https://github.com/pointfreeco/swift-composable-architecture/blob/47465c3f265cf058447c32c58a6676216ed76590/Sources/ComposableArchitecture/TestStore.swift#L800

https://github.com/pointfreeco/swift-composable-architecture/blob/47465c3f265cf058447c32c58a6676216ed76590/Sources/ComposableArchitecture/TestStore.swift#L1153


---

do you means  i should use `removeDuplicates` rather than `Equatable `to create `ViewState` ?
> Please note that an Equatable Reducer.State is not a requirement for ViewState. It's even sometimes an undesirable side effect of testability, as its immediate availability as Equatable encourages you to use it in a ViewStore when a dedicated (and Equatable) ViewState may be preferable.

**Reply by tgrapperon** [reply]
> do you means i should use removeDuplicates rather than Equatable to create ViewState ?

You shouldn't really use `removeDuplcates` in the general picture. What I'm saying is that if you have a `Parent.State`:
```swift
struct Parent {
  struct State: Equatable {
    var text: String
    var date: Date
	var child: Child.State
  }
}
```
and if you only need to access `text` and `date` in your `View`, you may be tempted to observe `Parent.State` in your `ViewStore` because it is convenient to do so, as this type is already `Equatable` (for testability reasons), and `observe: { $0 }` is only a few keystrokes away. But a `ViewStore` that observes all `Parent.State` will invalidate the view each time `child` changes, and this can create many issues. You should instead define and observe a custom:
```swift
struct ViewState: Equatable {
  let text: String
  let date: Date
  init(state: Parent.State) {
    self.text = state.text
    self.date = state.date
  }
}
```
There are of course legitimate cases where observing the whole `Reducer.State` makes sense, but we can't use non-equatability by default to force us to make `Reducer.State` `Equatable` explicitly when we want to observe them directly, as they already have to be `Equatable` for testability.
You can check the ["Performance" article](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/performance) for more informations.

**Reply by hmhv** [reply]
Thank you again.
I understand what you mean. `ViewState of minimum will be more performance`
